File: backend/src/main/java/com/plantit/entity/Complaint.java
package com.plantit.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.OffsetDateTime;

@Entity
@Table(name = "complaints")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Complaint {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "complaint_number", nullable = false, unique = true)
    private String complaintNumber;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "equipment_id")
    private Equipment equipment;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "location_id")
    private Location location;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reporter_id")
    private Employee reporter;

    /** The WhatsApp sender's display name, captured regardless of whether they match a known Employee - reporter above requires a phone-number match, this doesn't. */
    @Column(name = "reporter_name")
    private String reporterName;

    @Column(nullable = false)
    private String source;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "whatsapp_group_id")
    private WhatsAppGroup whatsappGroup;

    @Column(nullable = false)
    private String priority;

    @Column(nullable = false)
    private String status;

    @Column(name = "ai_confidence")
    private Double aiConfidence;

    @Column(name = "ai_category")
    private String aiCategory;

    @Column(name = "equipment_reference")
    private String equipmentReference;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @Column(name = "resolved_at")
    private OffsetDateTime resolvedAt;

    /** Who actually marked it resolved - the WhatsApp reply's sender, or the admin who corrected the status manually. Cleared alongside resolvedAt on a reopen. */
    @Column(name = "resolved_by")
    private String resolvedBy;

    @Column(name = "closed_at")
    private OffsetDateTime closedAt;
}

File: frontend/src/types/complaint.ts
export interface ComplaintSummary {
  id: number;
  complaintNumber: string;
  title: string;
  description?: string;
  source: string;
  priority: string;
  status: string;
  category?: string;
  equipment?: string;
  equipmentReference?: string;
  location?: string;
  department?: string;
  reporter?: string;
  resolvedBy?: string;
  group?: string;
  confidence?: number;
  createdAt: string;
  resolvedAt?: string;
}

export interface ComplaintEvent {
  eventType: string;
  oldValue?: string;
  newValue?: string;
  description?: string;
  performedBy?: string;
  createdAt: string;
}

export interface ComplaintDetail extends ComplaintSummary {
  categoryId?: number;
  equipmentId?: number;
  locationId?: number;
  imageUrls?: string[];
  events?: ComplaintEvent[];
}

export interface ReferenceOption {
  id: number;
  name: string;
  active: boolean;
}

export interface EquipmentOption extends ReferenceOption {
  equipmentCode: string;
}

export interface CategoryRecord extends ReferenceOption {
  description?: string | null;
}

export interface CategoryRequest {
  name: string;
  description: string | null;
  active: boolean;
}

export interface EquipmentRecord extends ReferenceOption {
  equipmentCode: string;
  type?: string | null;
}

export interface EquipmentRequest {
  equipmentCode: string;
  name: string;
  type: string | null;
  manufacturer: string | null;
  model: string | null;
  locationId: number | null;
  department: string | null;
  active: boolean;
}

/** Sent as-is on save - null on categoryId/equipmentId/locationId means "unassign". */
export interface ComplaintUpdatePayload {
  title: string;
  categoryId: number | null;
  equipmentId: number | null;
  equipmentReference: string | null;
  locationId: number | null;
  priority: string;
  status: string;
}
